Sources/Workspace/Workspace.swift

```diff
@@ -2561,168 +2684,208 @@ extension Workspace {
+            break
+        }
 
-                // Ensure that the alternative object store is still valid.
-                //
-                // This can become invalid if the build directory is moved.
-                guard workingCopy.isAlternateObjectStoreValid() else {
-                    break fetch
-                }
+        // Inform the delegate.
+        delegate?.removing(repository: dependency.packageRef.location)
 
-                // The fetch operation may update contents of the checkout, so
-                // we need do mutable-immutable dance.
-                try fileSystem.chmod(.userWritable, path: path, options: [.recursive, .onlyFiles])
-                try workingCopy.fetch()
-                try? fileSystem.chmod(.userUnWritable, path: path, options: [.recursive, .onlyFiles])
+        // Compute the dependency which we need to remove.
+        let dependencyToRemove: ManagedDependency
 
-                return path
-            }
-        }
-       
-        // If not, we need to get the repository from the checkouts.
-        let repository = try package.makeRepositorySpecifier()
-        // FIXME: this should not block
-        let handle = try temp_await {
-            repositoryManager.lookup(repository: repository, skipUpdate: true, on: .sharedConcurrent, completion: $0)
+        if case .edited(let _basedOn, let unmanagedPath) = dependency.state, let basedOn = _basedOn {
+            // Remove the underlying dependency for edited packages.
+            dependencyToRemove = basedOn
+            let updatedDependency = Workspace.ManagedDependency.edited(
+                packageRef: dependency.packageRef,
+                subpath: dependency.subpath,
+                basedOn: .none,
+                unmanagedPath: unmanagedPath
+            )
+            self.state.dependencies.add(updatedDependency)
+        } else {
+            dependencyToRemove = dependency
+            self.state.dependencies.remove(dependencyToRemove.packageRef.identity)
         }
 
-        // Clone the repository into the checkouts.
-        let path = self.location.repositoriesCheckoutsDirectory.appending(component: repository.basename)
-
-        try fileSystem.chmod(.userWritable, path: path, options: [.recursive, .onlyFiles])
-        try fileSystem.removeFileTree(path)
-
-        // Inform the delegate that we're starting cloning.
-        delegate?.willCreateWorkingCopy(repository: handle.repository.url, at: path)
-        _ = try handle.createWorkingCopy(at: path, editable: false)
-        delegate?.didCreateWorkingCopy(repository: handle.repository.url, at: path, error: nil)
+        switch package.kind {
+        case .root, .fileSystem:
+            fatalError("local dependencies are supported")
+        case .localSourceControl, .remoteSourceControl:
+            try self.removeRepository(dependency: dependencyToRemove)
+        case .registry:
+            fatalError("registry dependencies are supported at this point")
```

Comment:
@mattt this is where you would plug the registry remove handler, if one is needed (for cleanup)